### PR TITLE
Deprecate pvlib.forecast

### LINKED ISSUE #1057: deprecate existing code in forecast.py, possibly replace with solarforecastarbiter shim
`forecast.py` is a burden to maintain. I haven't used it in years, I don't think any of the other pvlib maintainers are interested in it, and I don't see any users stepping up to volunteer to maintain it. The code is not up to my present standards and I don't see how I'd get it there without a complete rewrite. This leads to difficult to track bugs such as the one recently reported on the [google group](https://groups.google.com/g/pvlib-python/c/b9HdgWV6w6g). It also complicates the pvlib dependencies.

[solarforecastarbiter](https://github.com/SolarArbiter/solarforecastarbiter-core) includes a [reference_forecasts](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/solarforecastarbiter/reference_forecasts) package that is much more robust. See [documentation here](https://solarforecastarbiter-core.readthedocs.io/en/latest/reference-forecasts.html) and [example notebook here](https://github.com/SolarArbiter/workshop/blob/master/reference_forecasts.ipynb) (no promises that this works without modification for the latest version).

The main reason to prefer `forecast.py` to `solarforecastarbiter` is the data fetch process. `forecast.py` pulls point data from a Unidata THREDDS server. `solarforecastarbiter.reference_forecasts` assumes you already have gridded data stored in a netcdf file. `solarforecastarbiter.io.nwp` provides functions to fetch that gridded data from NCEP. We have very good reasons for that approach in `solarforecastarbiter`, but I doubt that many `forecast.py` users are interested in configuring that two step process for their application.

I'm very tempted to stop here, remove `forecast.py` after deprecation, and say "not my problem anymore", but it seems to attract a fair number of people to pvlib, so I hesitate to remove it without some kind of replacement. Let's explore a few ideas.

1. Within `forecast.py`, rewrite code to fetch relevant data from Unidata. Make this function compatible with the expectations for the [`load_forecast`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/6200ec067bf83bc198a3af59da1d924d4124d4ec/solarforecastarbiter/reference_forecasts/models.py#L16-L19) function passed into `solarforecastarbiter.reference_forecasts.models` functions.
2. Same as 1., except put that code somewhere else. Could be a documentation example, could be in solarforecastarbiter, or could be in a gist.
3. Copy/refactor solarforecastarbiter code into `forecast.py`.
4. Do nothing and let the forecast.py bugs and technical debt pile up. 

Other thoughts?

### PR DESCRIPTION
**Pull request by kanderso-nrel**
- [x] Closes #1057
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: ci/azure/posix.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ jobs:
       versionSpec: '$(python.version)'
 
   - script: |
-      pip install pytest pytest-cov pytest-mock requests-mock pytest-timeout pytest-azurepipelines pytest-rerunfailures pytest-remotedata
+      pip install pytest==7.0.1 pytest-cov pytest-mock requests-mock pytest-timeout pytest-azurepipelines pytest-rerunfailures pytest-remotedata
```
Comment (kanderso-nrel) [inline ci/azure/posix.yml:26]: This fixes an unrelated issue.  Probably we'll be able to unpin pytest again once this is resolved: https://github.com/tonybaloney/pytest-azurepipelines/issues/57

File: docs/sphinx/source/user_guide/forecasts.rst [hunk 2]
```diff
@@ -4,6 +4,25 @@
 Forecasting
 ***********
 
+.. warning::
```
Comment (kanderso-nrel) [inline docs/sphinx/source/user_guide/forecasts.rst:7]: I suppose this information could go in the whatsnew as well.  Open to suggestions both for its location and its content.

**Review by kanderso-nrel — COMMENTED** [review]
Ready for review.  Some notes below.

File: ci/azure/posix.yml (hunk 1 above)
Comment (tonybaloney) [reply ci/azure/posix.yml:26]: Downstream issue is resolved.

Comment (kanderso-nrel) [reply ci/azure/posix.yml]: Thanks @tonybaloney!

**Review by wholmgren — APPROVED** [review]
ready to merge?

FWIW I remembered to check this PR after being annoyed at the doc build dependencies and time for unrelated changes.

File: docs/sphinx/source/user_guide/forecasts.rst (hunk 2 above)
Comment (wholmgren) [reply docs/sphinx/source/user_guide/forecasts.rst:7]: what's new links to here, so I think it's fine.

**Comment by cwhanse** [comment]
Do we intend to remove all of the forecast.py module, or perhaps keep a base layer of functions? That can be worked out later. 

I think it would help to add the warning to reference/forecasting.rst

File: docs/sphinx/source/whatsnew/v0.9.1.rst [hunk 3]
```diff
@@ -16,6 +16,8 @@ Deprecations
   :py:meth:`pvlib.modelchain.ModelChain.with_sapm` for alternative simplified
   :py:class:`~pvlib.modelchain.ModelChain` interfaces, although note that the
   inputs do not directly translate. (:pull:`1401`)
+* All functionality in the ``pvlib.forecasts`` module is deprecated.
```
Comment (kanderso-nrel) [inline docs/sphinx/source/whatsnew/v0.9.1.rst:19]:
```suggestion
* All functionality in the ``pvlib.forecast`` module is deprecated.
```

**Comment by kanderso-nrel** [comment]
> I think it would help to add the warning to reference/forecasting.rst

Good idea, done: https://pvlib-python--1426.org.readthedocs.build/en/1426/reference/forecasting.html#forecasting

**Comment by kanderso-nrel** [comment]
Thanks folks!

**Pull request merged**